zerver/lib/event_schema.py

```diff
@@ -1094,6 +1094,9 @@ def check_realm_user_update(
     )
 
 
+restart_event = event_dict_type(required_keys=[("type", Equals("restart")), ("immediate", bool)])
```

Comment:
Similar to one of my earlier comments, we don't want to add schema checkers of fixtures that we don't actually exercise in the tests.  Is there a meaningful test that we can do here?  I think it's some tornado-related code that sends the restart event, correct?  Even though the schema here is obviously pretty trivial, we should make sure it is what the actual code sends.

Reply:
Yeah, it's tornado code that sends the restart event, I think `zerver/tornado/autoreload.py` is the one that restart the server if any source file is changed. I'm not sure if we can add any test for it (I'll look into this). As for the response, we're missing `server_generation` in schema (I'll fix this).

Reply:
It seems pretty clear to me that `zerver/tornado/event_queue.py` has the code that actually send the event:

~~~
 565 def send_restart_events(immediate: bool = False) -> None:
 566     event: Dict[str, Any] = dict(type="restart", server_generation=settings.SERVER_GENERATION)
 567     if immediate:
 568         event["immediate"] = True
 569     for client in clients.values():
 570         if client.accepts_event(event):
 571             client.add_event(event)

Reply:
It also appears that `immediate` is technically an optional field.

Reply:
It would be nice if we had an existing test on that code, but unfortunately `event_queue.py` is one of the files where we don't have 100% coverage, and this function has zero coverage.

Fortunately, I think you can test this pretty simply by creating a mock version of `clients` to capture the events, and then just call `send_restart_events` in a test, see what the mock `clients` captured, and then compare them to the schema.

Reply:
Ok, I'm about to push a test to master here that you can modify.

Reply:
You can stack your PR on top of #17722 (or fetch the latest master if I have pushed the change to master by the team you read this).

Reply:
Going back to the `immediate` field, I said earlier that it's technically an optional field, but I think we should just tell a tiny lie in `event_schema.py` and say that it's mandatory.  It's an odd decision in the original code that we omit the field if it's False, instead of just setting it to False.  My test only exercises the True side of the condition.  If somebody has more expertise about the reload event than I do, then they can enhance that test after my changes and your changes and deal with it accordingly.  They can either tweak the schema to make it optional, or they can fix the code to use the explicit value.

Reply:
For the test that I wrote, I realized I didn't need to mock `clients`.  You can see my PR to see the strategy that I chose instead (I wanted a slightly deeper test).

Reply:
(#17722 has been merged to master)

Reply:
Thanks for the #17722. It should be good enough for our case.

The `immediate` was only meant for preventing the breaking of automatic reload of open browser window on server restart which was the result of dumping/loading in development for testing event queue dump/load, because we don't have any purpose for it in production that we decided to drop the field if it's False.

I was thinking of more end-to-end approach for the test though I'm still not sure about it (I'll discuss this on czo), so I might try to enhance this later.  Most probably we would be able to check for both the condition then.

Reply:
https://github.com/zulip/zulip/issues/15541 may also be relevant, though I wouldn't block this work on fixing that.